types/jest/index.d.ts

```diff
@@ -833,31 +818,19 @@ declare namespace jest {
          * Used when you want to check that an item is in a list.
          * For testing the items in the list, this uses `===`, a strict equality check.
          * It can also check whether a string is a substring of another string.
-         *
-         * Optionally, you can provide a type for the expected value via a generic.
-         * This is particularly useful for ensuring expected objects have the right structure.
          */
-        // tslint:disable-next-line: no-unnecessary-generics
-        toContain<E = any>(expected: E): R;
+        toContain(expected: T extends Array<infer E> ? E : T extends string ? string : never): R;
         /**
          * Used when you want to check that an item is in a list.
          * For testing the items in the list, this matcher recursively checks the
          * equality of all fields, rather than checking for object identity.
-         *
-         * Optionally, you can provide a type for the expected value via a generic.
-         * This is particularly useful for ensuring expected objects have the right structure.
          */
-        // tslint:disable-next-line: no-unnecessary-generics
-        toContainEqual<E = any>(expected: E): R;
+        toContainEqual(expected: T extends Array<infer E> ? E : never): R;
         /**
          * Used when you want to check that two objects have the same value.
          * This matcher recursively checks the equality of all fields, rather than checking for object identity.
-         *
-         * Optionally, you can provide a type for the expected value via a generic.
-         * This is particularly useful for ensuring expected objects have the right structure.
          */
-        // tslint:disable-next-line: no-unnecessary-generics
-        toEqual<E = any>(expected: E): R;
+        toEqual(expected: T): R;
```

Comment:
This is incorrect
This is a legitimate test
`expect( {a: undefined, b: 2} ).toEqual({b: 2});`
This change will have a typescript error

Reply:
I think this might still be an error after my changes.

It's a legitimate test but usually the left is not a literal but a return from a function for example and then either a is required and then it's more explicit (and correct) to also have it required on the left, or it's not required and then it will just keep working. 

I know we might break actual use cases but the gain on having this PR land is huge. E.g. if you refactor a field in an object, you used to have to manually update all your tests. Now these things will propagate because typescript will known that the expected value in toEqual is also of that type.

So yes, there is a small cost, by there is a huge win as well. I'd argue that the test you present is succeeding, but semantically the test doesn't make sense so it make sense to have a type error.

Reply:
As a quick though could this be used so that any explicit E is used else use T

Determining any from https://stackoverflow.com/questions/61624719/how-to-conditionally-detect-the-any-type-in-typescript
```
type UnionToIntersection<U> = 
  (U extends any ? (k: U)=>void : never) extends ((k: infer I)=>void) ? I : never

// If T is `any` a union of both side of the condition is returned.
type UnionForAny<T> = T extends never ? 'A' : 'B'

// Returns true if type is any, or false for any other type.
type IsStrictlyAny<T> =
  UnionToIntersection<UnionForAny<T>> extends never ? true : false

//---------------------------------------------------------------------------------------

//could this work ?
type ExplicitOrT<E,T> =  IsStrictlyAny<E> extends true ? T:E;
interface MyMatchers<T,R>{
    toEqual<E = any>(expected: ExplicitOrT<E,T>): R;
}

```

Reply:
> But what if updateSnapshotState was a function that mutated and deleted the some property.

You could always type expect instead of toEqual

```typescript
expect<MyType>(stuff).toEqual(otherStuff)
```

But if you allow deleting stuff, this should be reflected in your type. A lot of the examples you give are literals instead of using a type annotation (that is wider and actually covering the use case you want to cover). 

It might break users / tests but it's because the types are raised to be more correct and helping users to type their code better or fix their tests to match their type signatures.

The `ExplicitOrT` doesn't work. 

I can remove the undefined fields using something like the thing below though
```typescript
    type NonUndefinedKeys<T> = {
        [key in keyof T]: T[key] extends undefined ? never : key;
    }[keyof T];
    type NonUndefined<T> = {
        [key in NonUndefinedKeys<T>]: T[key]
    }
```

but I'd really argue that it would result in less safe types. 

```typescript
function add(x: number, y: number): number {
  return x + y
}
expect(add('2', '3')).toEqual('23')
```
also succeeds, does that mean that add should be typed as `function add(x: number | string, y: number | string): number | string {`? I'd argue not, as it's semantically a different function. 

Types add safety. 

And typescript choose to make a difference between `{foo: undefined}` and `{foo?: undefined}`, so I think typescript users would find it very weird that a type definition package doesn't follow that choice. At runtime, they are more or less the same, but still typescript choose to make it a difference. So if users type something as `type Foo = {foo: undefined}` they force themselves to always provide a `foo` when they create a `Foo`. They can always redefine their type as `type Foo = {foo?: undefined}`. 

If I type something as `type Foo = {foo: undefined | number}` for example, I want to make sure, I always pass a foo. So I think it's an important part of the signature so I want to force myself to also add it during tests to make it explicit foo is undefined for this case. I'd really hate it if @types/jest doesn't give me the safety I deserve

Reply:
> The ExplicitOrT doesn't work.

Correct - it was just a very quick thought.

> You could always type expect instead of toEqual
expect<MyType>(stuff).toEqual(otherStuff)

How is that supposed to help ?

> function add(x: number, y: number): number {
  return x + y
}
expect(add('2', '3')).toEqual('23')
also succeeds,

add takes two numbers so passing two strings is not going to succeed !

I expect that what you mean is that expect(5).toEqual('23') is permitted.
Ok, so you have not found a way to type toEqual correctly.
Returning to what you said before.

> I know we might break actual use cases 

> I'd argue that the test you present is succeeding, but semantically the test doesn't make sense so it make sense to have a type error.

The point is there is only one use case for toEqual and that is its behaviour.  If typescript prohibits that then it is not correctly typed.  

To further add to the problems with this pull request 

This test is now flagged as an error.
```
expect(true).not.toEqual('wtf')
```

Additionally, asymmetric matchers are now not permitted and these are permittable inputs to 
toEqual
toStrictEqual
toMatchObject  ( which you have given up on )
and all of the spy matchers
```
const fooOrBar = {
  asymmetricMatch: actual => actual === 'foo'  || actual === 'bar'
}
expect('foo').toEqual(fooOrBar)

```

> Types add safety.

>  I'd really hate it if @types/jest doesn't give me the safety I deserve

I'd say that currently the definitions are type safe.  The types that are allowed in are acceptable inputs.

Reply:
Further to above
You can look at the last time someone tried to attempt what you are trying to achieve https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020

I think it is highly unlikely that this pull request will be merged.

Reply:
>You could always type expect instead of toEqual
> expect(stuff).toEqual(otherStuff)

I meant 

```
expect<Type>(stuff).toEqual(otherStuff)
```

>I expect that what you mean is that expect(5).toEqual('23') is permitted.
Ok, so you have not found a way to type toEqual correctly.
Returning to what you said before.

No, I meant that even though `add('2', '3')` does something, it has a different semantic meaning and typing it as `(number, number) => number` makes sense. I'd argue with `expect/toEqual` it's the same. 

```
expect(true).not.toEqual('wtf')
expect(myBooleanReturningFunction()).not.toEqual('wtf')
```
Adds nothing if you have types. If `myBooleanReturningFunction` returns a boolean, why even check if it doesn't equal a string. You know it won't do this because Typescript tells you to. You probably made a mistake so it makes sense to have a type error there. You can always type it as 

```
expect<boolean|string>(myBooleanReturningFunction()).not.toEqual('wtf')
```
if you do want to type your tests and have this test. Or not have your tests typed. But if you have types in your tests, why not actually have types in your test? 

>Most importantly, asymmetric matchers are now not permitted and these are permittable inputs

If we are missing something, let's add them. But I don't think that's an argument against this PR. It's something we just need to fix (and shouldn't be very hard)

> I think it is highly unlikely that this pull request will be merged.
Why are you not leading with this then? Then I can just fork it and actually have type safety. I wanted to contribute back to the community so everyone can enjoy more type-safe tests. I know this will result in type errors for people because the current types are not very safe and so it will raise some genuine issues with tests that evolved and that people didn't realize.

Reply:
> the current types are not very safe 
They are perfectly type safe.  

> I know this will result in type errors for people

Are you aware that breaking changes are not permitted on definitely typed until there is an accompanying major release of the javascript library ?  

> Why are you not leading with this then?
There is a difference between this pull request and the previous.  The previous change was not for a major version change in jest.  Regardless, there will definitely be opposition but perhaps you can convince people that it should be accepted.  I know you certainly will try.  

I am looking to minimise these breaking changes and to correct the mistakes you have made in the context of your argument.  You are so sure that T -> T yet you are not fully aware of the behaviour of the functions that you are typing, let alone how other developers legitimately use them differently.

All of this aside there is one way you could have what you want without impacting anyone else.
You could introduce a new type that mirrors jest.Expect but with the typing that you are interested in - e.g jest.ExpectT.  Then in your test you redefine the type of expect.

This is not something I would object to although others may.

Reply:
> They are perfectly type safe.

Well you could have a test
```typescript
expect(myMockFn).not.toHaveBeenCalledWith(any, data)
```
and then change the signature of `myMockFn` and all of a sudden this test will always pass because the test is not even following the signature anymore. The test is useless and typescript could prevent that but the current types are not type safe so they don't catch this. 

> Are you aware that breaking changes are not permitted on definitely typed until there is an accompanying major release of the javascript library ?

No, I was not, but we can try to make them non breaking. The type errors are mostly there because people are testing bogus things. It's logical that on an upgrade, you could get type errors. 

> I am looking to minimise these breaking changes and to correct the mistakes you have made in the context of your argument. You are so sure that T -> T yet you are not fully aware of the behaviour of the functions that you are typing, let alone how other developers legitimately use them differently.

I want to have the same! I want to make it as little disruptive as possible, that's why I moved back to the `extends T` approach, so we could have type safety without breaking too many people. 

And yes, I might not be covering all use cases, so I'm happy to do the work to have them, if it results in this getting merged. I don't want to do the work if I have to fork it anyway. 

I assumed that the tests were how people used jest and if I got those to pass, then I'd be supporting everything, if there is not great test coverage, let's try to increase it and I'll try to make them work with this change!

> You could introduce a new type that mirrors jest.Expect but with the typing that you are interested in - e.g jest.ExpectT. Then in your test you redefine the type of expect.

That sounds like it's almost as much work as just forking the repo? 

> All of this aside, there is one way you could have what you want without impacting anyone else.
My big concern is that we hold on to the way jest typings used to be instead of looking to what they can be. I know for existing projects it's not ideal, but potentially as a major version it could make sense? So newer projects can start with more safety in place?

Reply:
> typescript could prevent that but the current types are not type safe 

Yes typescript can prevent that.  This is the final comment where I reiterate - the current types are type safe ( certainly with respect to your argument of T.  The mock/spy matchers could be changed to enforce actual to be a mock/spy ).  They describe the permittable argument types.  Just because the type definitions are set up so that TActual has been made available to matchers does not mean that the methods are generic.
If a developer wants the behaviour that you are going to enforce on everyone they can use typescript to do so themselves.  Yes the current helper generic is not sufficient for your test typing needs but it is the only solution that does not narrow the permittable argument types and arguably making the typings incorrect.

I have to ask.  Have you read and understood the counter arguments in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020 and mentioned threads?

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-740569315
Not sure if this is still valid https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-738288238
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47553#issuecomment-692270723

Another example is the use of fake types although I cannot find the comment currently.

> I assumed that the tests were how people used jest and if I got those to pass, then I'd be supporting everything, if there is not great test coverage, let's try to increase it and I'll try to make them work with this change!

Of course the tests are being used for testing but in ways that you did not consider and will no longer be permittable without a cast.  

> I know for existing projects it's not ideal, but potentially as a major version it could make sense?

Perhaps.  There is no doubt that the argument for is strong.

> ...happy to do the work to have them, if it results in this getting merged. I don't want to do the work if I have to fork it anyway.

Well it certainly isn't going to be merged if you do not do the remaining work. In my opinion you still need to type toEqual so that it behaves as to toEqual.  Also, even when you have perfectly typed the changes that you want to make and there is sufficient tests for these changes and it is agreed that the new typing is preferable to the old, it will not be merged until a new version of jest is released, as it is a breaking change.

> You could introduce a new type that mirrors jest.Expect but with the typing that you are interested in - e.g jest.ExpectT. Then in your test you redefine the type of expect.
That sounds like it's almost as much work as just forking the repo?

I thought you wanted to contribute back to the community ? If you did this ( and it would not be that much work ) then immediately people could use this change that is so beneficial ( of course making them aware of its existence is a problem ).  It could also be the route to getting the change to be the permanent default.

An alternative solution, albeit not as strong, is to use eslint.  At least with that it is an opt in.




linter